File: VB/MyXtraGrid/MyGridView.vb
Imports System
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraEditors.Helpers
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Collections.Generic

Namespace MyXtraGrid

    Public Class MyGridView
        Inherits DevExpress.XtraGrid.Views.Grid.GridView

        Public Sub New()
            Me.New(Nothing)
        End Sub

        Public Sub New(ByVal grid As DevExpress.XtraGrid.GridControl)
            MyBase.New(grid)
        End Sub

        Protected Overrides ReadOnly Property ViewName As String
            Get
                Return "MyGridView"
            End Get
        End Property

        Protected Overrides Function CreateDateFilterPopup(ByVal column As GridColumn, ByVal ownerControl As Control, ByVal creator As Object) As DateFilterPopup
            Return New MyDateFilterPopup(Me, column, ownerControl, creator)
        End Function
    End Class

    Public Class MyDateFilterPopup
        Inherits DateFilterPopup

        Public Sub New(ByVal view As ColumnView, ByVal column As GridColumn, ByVal ownerControl As Control, ByVal creator As Object)
            MyBase.New(view, column, ownerControl, creator)
        End Sub

        Private dateCalendarField As DateControlEx = Nothing

        Private ReadOnly Property DateCalendar As DateControlEx
            Get
                If dateCalendarField Is Nothing Then SetDateCalendar(DateFilterControl)
                Return dateCalendarField
            End Get
        End Property

        Private dateFilterControlField As PopupOutlookDateFilterControl = Nothing

        Private ReadOnly Property DateFilterControl As PopupOutlookDateFilterControl
            Get
                If dateFilterControlField Is Nothing Then SetDateFilterControl(item)
                Return dateFilterControlField
            End Get
        End Property

        Private item As RepositoryItemPopupContainerEdit = Nothing

        Protected Overrides Function CreateRepositoryItem() As RepositoryItemPopupBase
            item = TryCast(MyBase.CreateRepositoryItem(), RepositoryItemPopupContainerEdit)
            AddHandler DateCalendar.CustomDrawDayNumberCell, AddressOf OnCustomDrawDayNumberCell
            Return item
        End Function

        Private Sub SetDateFilterControl(ByVal item As RepositoryItemPopupContainerEdit)
            For Each ctrl As Control In item.PopupControl.Controls
                If TypeOf ctrl Is PopupOutlookDateFilterControl Then
                    dateFilterControlField = TryCast(ctrl, PopupOutlookDateFilterControl)
                    Exit For
                End If
            Next
        End Sub

        Private Sub SetDateCalendar(ByVal dateFilterControl As PopupOutlookDateFilterControl)
            For Each c As Control In dateFilterControl.Controls
                If TypeOf c Is DateControlEx Then
                    dateCalendarField = TryCast(c, DateControlEx)
                    Exit For
                End If
            Next
        End Sub

        Protected Overloads ReadOnly Property View As MyGridView
            Get
                Return TryCast(MyBase.View, MyGridView)
            End Get
        End Property

        Private uniqueFilterPopupValuesField As Object() = Nothing

        Private ReadOnly Property UniqueFilterPopupValues As Object()
            Get
                If uniqueFilterPopupValuesField Is Nothing Then uniqueFilterPopupValuesField = GetUniqueFilterPopupValues()
                Return uniqueFilterPopupValuesField
            End Get
        End Property

        Private Function GetUniqueFilterPopupValues() As Object()
            Dim dates As List(Of Object) = New List(Of Object)()
            For i As Integer = 0 To View.DataRowCount - 1
                Dim val As Object = View.GetRowCellValue(i, Column)
                If Not dates.Contains(val) Then dates.Add(val)
            Next

            Return dates.ToArray()
        End Function

        Private Sub OnCustomDrawDayNumberCell(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Calendar.CustomDrawDayNumberCellEventArgs)
            If IsExistingDate(e.Date) Then e.Style.ForeColor = Color.BlueViolet
        End Sub

        Private Function IsExistingDate(ByVal dt As Date) As Boolean
            Dim index As Integer = Array.IndexOf(UniqueFilterPopupValues, dt)
            Return index <> -1
        End Function

        Public Overrides Sub Dispose()
            If DateCalendar IsNot Nothing Then RemoveHandler DateCalendar.CustomDrawDayNumberCell, AddressOf OnCustomDrawDayNumberCell
            MyBase.Dispose()
        End Sub
    End Class
End Namespace

File: CS/MyXtraGrid/MyGridView.cs

using System;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraEditors.Helpers;
using System.Windows.Forms;
using System.Drawing;
using DevExpress.Utils;
using DevExpress.Data;
using System.Collections.Generic;

namespace MyXtraGrid {
	public class MyGridView : DevExpress.XtraGrid.Views.Grid.GridView {
		public MyGridView() : this(null) {}
		public MyGridView(DevExpress.XtraGrid.GridControl grid) : base(grid) {
		}
		protected override string ViewName { get { return "MyGridView"; } }

        protected override DateFilterPopup CreateDateFilterPopup(DevExpress.XtraGrid.Columns.GridColumn column, System.Windows.Forms.Control ownerControl, object creator) {
            return new MyDateFilterPopup(this, column, ownerControl, creator);
        }
	}

    public class MyDateFilterPopup: DateFilterPopup {

        public MyDateFilterPopup(ColumnView view, GridColumn column, System.Windows.Forms.Control ownerControl, object creator) : base(view, column, ownerControl, creator) { }
        
        DateControlEx dateCalendar = null;
        DateControlEx DateCalendar {
            get {
                if(dateCalendar == null) SetDateCalendar(DateFilterControl);
                return dateCalendar;
            }
        }
        PopupOutlookDateFilterControl dateFilterControl = null;
        PopupOutlookDateFilterControl DateFilterControl {
            get {
                if(dateFilterControl == null) SetDateFilterControl(item);
                return dateFilterControl;
            }
        }
        RepositoryItemPopupContainerEdit item = null;

        protected override RepositoryItemPopupBase CreateRepositoryItem() {
            item = base.CreateRepositoryItem() as RepositoryItemPopupContainerEdit;
            DateCalendar.CustomDrawDayNumberCell += OnCustomDrawDayNumberCell;
            return item;
        }

        private void SetDateFilterControl(RepositoryItemPopupContainerEdit item) {
            foreach(Control ctrl in item.PopupControl.Controls)
                if(ctrl is PopupOutlookDateFilterControl) {
                    dateFilterControl = ctrl as PopupOutlookDateFilterControl;
                    break;
                }
        }

        private void SetDateCalendar(PopupOutlookDateFilterControl dateFilterControl) {
            foreach(Control c in dateFilterControl.Controls)
                if(c is DateControlEx) {
                    dateCalendar = c as DateControlEx;
                    break;
                }
        }

        protected new MyGridView View {
            get { return base.View as MyGridView; }
        }

        object[] uniqueFilterPopupValues = null;
        object[] UniqueFilterPopupValues {
            get {
                if(uniqueFilterPopupValues == null)
                    uniqueFilterPopupValues = GetUniqueFilterPopupValues();
                return uniqueFilterPopupValues; }
        }

        object[] GetUniqueFilterPopupValues() {
            List<object> dates = new List<object>();
            for(int i = 0;i < View.DataRowCount;i++) {
                object val = View.GetRowCellValue(i, Column);
                if(!dates.Contains(val))
                    dates.Add(val);
            }
            return dates.ToArray();
        }

        void OnCustomDrawDayNumberCell(object sender, DevExpress.XtraEditors.Calendar.CustomDrawDayNumberCellEventArgs e) {
            if(IsExistingDate(e.Date))
                e.Style.ForeColor = Color.BlueViolet;
        }

        bool IsExistingDate(DateTime dt) {
            int index = Array.IndexOf(UniqueFilterPopupValues, dt);
            return index != -1;
        }

        public override void Dispose() {
            if(DateCalendar != null)
                DateCalendar.CustomDrawDayNumberCell -= OnCustomDrawDayNumberCell;
            base.Dispose();
        }
    }
}
